testutils: check all ServerArgs for compatibility with a TestCluster

andreimatei opened the pull request:
<!-- Reviewable:start -->

This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/9600)

<!-- Reviewable:end -->

andreimatei committed:
testutils: check all ServerArgs for compatibility with a TestCluster

Before this patch, only TestCluster.ServerArgs were checked, not
TestCluster.ServerArgsPerNode.

Also remove unnecessary initialization from a gossip test.

BramGruneir commented on testutils/testcluster/testcluster.go:
```diff
@@ -122,31 +122,40 @@ func StartTestCluster(t testing.TB, nodes int, args base.TestClusterArgs) *TestC
-		t.Fatal("can't disable an individual server's queues when starting a cluster; " +
-			"the cluster controls replication")
+
+	// Sanity-check and massage all ServerArgs to work for a cluster.
```
What do you mean by massage here?

andreimatei replied:
you know... "fiddle with". "to work for a cluster" is the more suggestive part of the comment.

BramGruneir replied:
Yeah, it's just ambiguous.  Could you elaborate on what you're actually doing?  
I know it's mostly the sanity check, so perhaps just remove "massage" entirely.

tamird commented:
Reviewed 1 of 1 files at r1.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-KSlvk_IZfRZZF8TyRLq:b-6hofl9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
> ```Go
> 	allArgs = append(allArgs, &args.ServerArgs)
> 	for _, sargs := range args.ServerArgsPerNode {
> 		allArgs = append(allArgs, &sargs)
> ```

you're taking a pointer to a `for-range` binding. :sheep: 

why is `allArgs` a slice of pointers, anyway?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
LGTM with the slight change in wording and addressing @tamird's issue

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions.

---

*[testutils/testcluster/testcluster.go, line 126 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlp0Ybdu-yyEXurJ6o-r1-126:-KTpMCdgRGzF-kWjbh5h:b-pmzgnp) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L126)):*
<details><summary><i>Previously, BramGruneir (Bram Gruneir) wrote…</i></summary>
> Yeah, it's just ambiguous.  Could you elaborate on what you're actually doing?  
> 
> I know it's mostly the sanity check, so perhaps just remove "massage" entirely.
</details>
i said "initialize"

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-KTpMEVa68ln9yoovi7G:bp5cuzx) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary>
> you're taking a pointer to a `for-range` binding. :sheep: 
> 
> why is `allArgs` a slice of pointers, anyway?
</details>
oops good catch!
they're pointers cause I need to change `sargs.Knobs.Store` below.
See what I'm doing now; hopefully I'm being silly and there's a better way...

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-KTpTe6ubrNXhIH41pOz:b-qacb77) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary>
> oops good catch!
> they're pointers cause I need to change `sargs.Knobs.Store` below.
> See what I'm doing now; hopefully I'm being silly and there's a better way...
</details>
Hm. I haven't the faintest idea what this code is trying to do. Can you elaborate on what kind of sanity check you're doing here? What is the point of this map? What is the point of the `-1` key? Nothing is using it.

So lost.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-KTtcsT2etZrwDm3wp1m:b-lnkph) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
Well I was hoping this would explain the intent:
> // To treat the default args and the per-server args uniformly, we'll copy
> 	// them into a map of pointers and the copy them back.

So the idea is that we want to go through both `args.ServerArgs` and `args.ServerArgsPerNode` (a map) and apply some transformation to both of them. So I copy both of them to a collection of pointers, manipulate them, then copy them back.
The `-1`is used at the `// Copy the possibly-modified args back into `args`.` step, where we only copy keys >= 0. Maybe I should make that more explicit.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

RaduBerinde commented:
Is this still needed? If not, please close it..

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-KXkRiFZA5SgiX_l-m5h:b-fa0owc) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Well I was hoping this would explain the intent:
> // To treat the default args and the per-server args uniformly, we'll copy
> 	// them into a map of pointers and the copy them back.

So the idea is that we want to go through both `args.ServerArgs` and `args.ServerArgsPerNode` (a map) and apply some transformation to both of them. So I copy both of them to a collection of pointers, manipulate them, then copy them back.
The `-1`is used at the `// Copy the possibly-modified args back into `args`.` step, where we only copy keys >= 0. Maybe I should make that more explicit.

</blockquote></details>

This is pretty damn complicated for what it's doing. I would just make what's inside the loop into a function and just call it once with `&args.ServerArgs` and then in a loop with `&args.ServerArgsPerNode[i]`.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KXkSq0FYk9THCLauRCs:blafgfa)*
<!-- Sent from Reviewable.io -->

BramGruneir commented:
@andreimatei What's going on with this PR?

andreimatei commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-K_1UY_8u5VGK7V0XN0R:b3yrbh9) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

This is pretty damn complicated for what it's doing. I would just make what's inside the loop into a function and just call it once with `&args.ServerArgs` and then in a loop with `&args.ServerArgsPerNode[i]`.
</blockquote></details>

well that doesn't quite work because in Go `&args.ServerArgsPerNode[i]` is not a thing. But I've extracted the body of the loop into a function.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r3.
Review status: 0 of 1 files reviewed at latest revision, 2 unresolved discussions, all commit checks successful.

---

*[testutils/testcluster/testcluster.go, line 130 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KSlvk_IZfRZZF8TyRLp:-K_7FxXn0GQgfAfVhe3K:bmbdsqd) ([raw file](https://github.com/cockroachdb/cockroach/blob/ebd1bcc7fd3993fc0a2817608e778cac1748f960/testutils/testcluster/testcluster.go#L130)):*
> in Go &args.ServerArgsPerNode[i] is not a thing.

What do you mean? I'm pretty sure it's a thing.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 2 files at r3.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/testutils/testcluster/testcluster.go, line 131 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7G_FE2bFWlyioZbLb:-K_7G_FE2bFWlyioZbLc:btk7952) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L131)):*
> ```Go
> 		t.Fatal(err)
> 	}
> 	for nodeIdx, sargs := range args.ServerArgsPerNode {
> ```

I'm pretty sure you don't need `sargs` here. You can write this more simply as:
```
for i := range args.ServerArgsPerNode {
  if err := initServerArgsForCluster(&args.ServerArgsPerNode[i], args.ReplicationMode); err != nil {
    t.Fatal(err)
  }
}
```

---

*[pkg/testutils/testcluster/testcluster.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7GvNkT8isBbmEh8pL:-K_7GvNkT8isBbmEh8pM:b-pyz4bo) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L177)):*
> ```Go
> ) error {
> 	if args.JoinAddr != "" {
> 		return errors.Errorf("can't specify a join addr when starting a cluster")
> ```

this error should include args.JoinAddr

---

*[pkg/testutils/testcluster/testcluster.go, line 184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HAtQN49-6nTmWfbB:-K_7HAtQN49-6nTmWfbC:b-f9l1e4) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L184)):*
> storeKnobs.(*storage.StoreTestingKnobs)
 
extract local plz

also, this seems fragile if/when we add the ability to disable other queues.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-K_7HNTjYdPVP_rm_wv-:beuh5n7) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
> ```Go
> 		storeKnobs.DisableReplicateQueue = true
> 	default:
> 		return errors.Errorf("unexpected replication mode")
> ```

include the "unexpected" replication mode in this error plz.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 0 of 1 files reviewed at latest revision, 6 unresolved discussions, some commit checks pending.

---

*[pkg/testutils/testcluster/testcluster.go, line 131 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7G_FE2bFWlyioZbLb:-K_BG9x2Q1pwCUMt5aj6:b38m1qm) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L131)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I'm pretty sure you don't need `sargs` here. You can write this more simply as:
```
for i := range args.ServerArgsPerNode {
  if err := initServerArgsForCluster(&args.ServerArgsPerNode[i], args.ReplicationMode); err != nil {
    t.Fatal(err)
  }
}
```
</blockquote></details>

`ServerArgsPerNode` is a map; you can't take the address of an element of the map :(

---

*[pkg/testutils/testcluster/testcluster.go, line 177 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7GvNkT8isBbmEh8pL:-K_BG_D5TyqmkxRlgLak:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L177)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this error should include args.JoinAddr
</blockquote></details>

Done.

---

*[pkg/testutils/testcluster/testcluster.go, line 184 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HAtQN49-6nTmWfbB:-K_BHSF7GriL-lhy08QL:b8c6fc8) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L184)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> storeKnobs.(*storage.StoreTestingKnobs)
 
extract local plz

also, this seems fragile if/when we add the ability to disable other queues.
</blockquote></details>

Done.
I've added a comment below to keep the queues we disable in sync with this check.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-K_BGaIx3iLVjn2WcoxK:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

include the "unexpected" replication mode in this error plz.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
This PR has been reduced in scope; now it's just about some sanity checks.
PTAL

cc @a-robinson for the gossip test change.

---

Review status: 0 of 8 files reviewed at latest revision, 6 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjPW4WhHFqgyr4uP_-R:b-85cmio)*
<!-- Sent from Reviewable.io -->

tamird commented:
Haven't review the 1st commit in detail, but I'm afraid I'm strongly :-1: on the 2nd. I think that change is way more controversial than the 1st commit; could you send a separate PR for that?

---

Reviewed 2 of 3 files at r4.
Review status: 1 of 8 files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjPhBDipmvwrCI6fvfe:b-zfbb38)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Well if I can't get you on board on the 2nd commit, I probably don't have much chances. I've dropped it.
But I thought you'd support the agenda of passing `testing.T` less around and returning errors from more testutils functions.

---

Review status: 2 of 3 files reviewed at latest revision, 4 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjPxsDmzzn0Ms00q0PB:bg6ugqh)*
<!-- Sent from Reviewable.io -->

tamird commented:
Well I would support that goal, but that solution is too clever for my
palate. I would prefer to see the explicit refactoring.

I'll review this later tonight or tomorrow!

On May 5, 2017 19:36, "Andrei Matei" <notifications@github.com> wrote:

> Well if I can't get you on board on the 2nd commit, I probably don't have
> much chances. I've dropped it.
> But I thought you'd support the agenda of passing testing.T less around
> and returning errors from more testutils functions.
> ------------------------------
>
> Review status: 2 of 3 files reviewed at latest revision, 4 unresolved
> discussions.
> ------------------------------
>
> *Comments from Reviewable
> <https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjPxsDmzzn0Ms00q0PB:bg6ugqh>*
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9600#issuecomment-299598618>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPBoIJ_VDEdZvWa0-cCqeUCQgzNQnks5r27KZgaJpZM4KJFvY>
> .
>

andreimatei commented:
No rush!

On May 5, 2017 8:01 PM, "Tamir Duberstein" <notifications@github.com> wrote:

> Well I would support that goal, but that solution is too clever for my
> palate. I would prefer to see the explicit refactoring.
>
> I'll review this later tonight or tomorrow!
>
> On May 5, 2017 19:36, "Andrei Matei" <notifications@github.com> wrote:
>
> > Well if I can't get you on board on the 2nd commit, I probably don't have
> > much chances. I've dropped it.
> > But I thought you'd support the agenda of passing testing.T less around
> > and returning errors from more testutils functions.
> > ------------------------------
> >
> > Review status: 2 of 3 files reviewed at latest revision, 4 unresolved
> > discussions.
> > ------------------------------
> >
> > *Comments from Reviewable
> > <https://reviewable.io:443/reviews/cockroachdb/cockroach/
> 9600#-:-KjPxsDmzzn0Ms00q0PB:bg6ugqh>*
> >
> > —
> > You are receiving this because you were mentioned.
> > Reply to this email directly, view it on GitHub
> > <https://github.com/cockroachdb/cockroach/pull/
> 9600#issuecomment-299598618>,
> > or mute the thread
> > <https://github.com/notifications/unsubscribe-auth/ABdsPBoIJ_VDEdZvWa0-
> cCqeUCQgzNQnks5r27KZgaJpZM4KJFvY>
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/9600#issuecomment-299600880>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAXBcSGnF4p2ZZP8H5UrdFPsWzVlEPD1ks5r27g1gaJpZM4KJFvY>
> .
>

tamird commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r5.
Review status: all files reviewed at latest revision, 7 unresolved discussions, all commit checks successful.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-KjPgiKOhAPPMV0NnEov:bcylbgh) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Done.
</blockquote></details>

doesn't look done. while you're changing this, it might look better as a `switch`.

---

*[pkg/testutils/testcluster/testcluster.go, line 136 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYe3oK2w8-DXFA43r_:-KjYe3oK2w8-DXFA43ra:b-on1g2f) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L136)):*
> ```Go
> 	}
> 	for _, sargs := range args.ServerArgsPerNode {
> 		err := checkServerArgsForCluster(sargs, args.ReplicationMode, disallowJoinAddr)
> ```

nit: why not use the same `if err := ...; err != nil` form you use above?

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-KjYfCg106LJZZjVsowS:bwpsr0c) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
> ```Go
> }
> 
> type checkType int
> ```

might just be simple as a boolean, but I don't feel strongly.

---

*[pkg/testutils/testcluster/testcluster.go, line 185 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYeDb8XSgNVNjwHPvI:-KjYeDb8XSgNVNjwHPvJ:bn4fpsj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L185)):*
> ServerArgs

Should this be "args"? There's nothing called ServerArgs.

---

*[pkg/testutils/testcluster/testcluster.go, line 246 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYf4eDuu6yJTul1f46:-KjYf4eDuu6yJTul1f47:b-1b2ntv) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L246)):*
> ```Go
> 	}
> 
> 	s, conn, _ := serverutils.StartServer(t, serverArgs)
> ```

so close to not needing `t` here at all :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjYf8GNuw2wpHVhBdSG:bnfp4nl)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 3 files reviewed at latest revision, 7 unresolved discussions.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-KjdHgCg5MMOcknZJ-fF:b-uya92o) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

doesn't look done. while you're changing this, it might look better as a `switch`.
</blockquote></details>

Sorry, modified the err message.
I liked the switch originally too, but now that it's been changed, I'll leave it.

---

*[pkg/testutils/testcluster/testcluster.go, line 136 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYe3oK2w8-DXFA43r_:-KjdHEcGPgr7omwCvmv0:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L136)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: why not use the same `if err := ...; err != nil` form you use above?
</blockquote></details>

Done.

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-KjdHzLvnDqRFFNmESfm:b5l1svi) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

might just be simple as a boolean, but I don't feel strongly.
</blockquote></details>

I want to experiment with more of these named constants for such args, as I've really come to dislike the proliferation of unreadable arguments to function calls. And people don't always care to comment them inline.

---

*[pkg/testutils/testcluster/testcluster.go, line 185 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYeDb8XSgNVNjwHPvI:-KjdHQi6PCaMGoQI9gqg:bh3sng5) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L185)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> ServerArgs

Should this be "args"? There's nothing called ServerArgs.
</blockquote></details>

the type it's checking is called `base.TestServerArgs`

---

*[pkg/testutils/testcluster/testcluster.go, line 246 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYf4eDuu6yJTul1f46:-KjdHp-D96YgInW7TJdU:ba4qmvj) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L246)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

so close to not needing `t` here at all :(
</blockquote></details>

not really close. The `t` is needed for some `pgurl` functions that are used in a ton of places.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks pending.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-KjdJjYNYNA7XvNhwTcx:bgpu1gl) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

Sorry, modified the err message.
I liked the switch originally too, but now that it's been changed, I'll leave it.
</blockquote></details>

Does ReplicationMode not implement fmt.Stringer? It'd be nice to %s it here, if possible.

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-KjdJG7y1Xv5MY1yH853:b-v3kmzw) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

I want to experiment with more of these named constants for such args, as I've really come to dislike the proliferation of unreadable arguments to function calls. And people don't always care to comment them inline.
</blockquote></details>

Understood, but this one is binary by definition. Perhaps it'd be clearer as

```
type checkType boolean

const (
  joinAddrAllow = true
  joinAddrDisallow = false
)
```

It's also unfortunate that this is inconsistent with `initialClusterStart`, the other boolean in this PR.

---

*[pkg/testutils/testcluster/testcluster.go, line 185 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYeDb8XSgNVNjwHPvI:-KjdJe8KQM6Zr5cdQWPN:b-4xwm9y) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L185)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

the type it's checking is called `base.TestServerArgs`
</blockquote></details>

Yes, that's my point, that type is not called ServerArgs.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 1 of 4 files reviewed at latest revision, 5 unresolved discussions.

---

*[pkg/testutils/testcluster/testcluster.go, line 200 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-K_7HNTjYdPVP_rm_wuz:-KjdYS55zQkL19_EI94I:bcvv55) ([raw file](https://github.com/cockroachdb/cockroach/blob/133f36a72e2a97ad5321e8c0d405366e8dd58f7b/pkg/testutils/testcluster/testcluster.go#L200)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Does ReplicationMode not implement fmt.Stringer? It'd be nice to %s it here, if possible.
</blockquote></details>

ok, added a `go:generate stringer` for it.

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-KjdWOV_FYH1vmtqecCb:b-pxj560) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Understood, but this one is binary by definition. Perhaps it'd be clearer as

```
type checkType boolean

const (
  joinAddrAllow = true
  joinAddrDisallow = false
)
```

It's also unfortunate that this is inconsistent with `initialClusterStart`, the other boolean in this PR.
</blockquote></details>

ah, `bool` is probably better. Switched to it.

---

*[pkg/testutils/testcluster/testcluster.go, line 185 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYeDb8XSgNVNjwHPvI:-KjdWY14K04ZQdULi3sF:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L185)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Yes, that's my point, that type is not called ServerArgs.
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 3 of 3 files at r7.
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/base/test_server_args.go, line 118 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdZGCEfx-9Z2IgLy3Q:-KjdZGCEfx-9Z2IgLy3R:bmi8ppe) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e8d6d3a460835e09537efa6cc46a2ea5c34dc03/pkg/base/test_server_args.go#L118)):*
> ```Go
> type TestClusterReplicationMode int
> 
> //go:generate stringer -type=TestClusterReplicationMode
> ```

:100: 

---

*[pkg/testutils/testcluster/testcluster.go, line 183 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdZKLUeNO93DVWfEB-:-KjdZKLUeNO93DVWfEB0:byhgz28) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e8d6d3a460835e09537efa6cc46a2ea5c34dc03/pkg/testutils/testcluster/testcluster.go#L183)):*
> ```Go
> const (
> 	disallowJoinAddr checkType = false
> 	allowJoinAddr              = true
> ```

Sadly, you need the `checkType` type annotation on both of these, otherwise you get an untyped constant. See https://play.golang.org/p/TsChFarg3g :(

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions.

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-Kjd_4gE-hK1V0uZJGqy:b-czuic9) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

ah, `bool` is probably better. Switched to it.
</blockquote></details>

Oh, there is also the inconsistency with initialClusterStart - can you address that?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: all files reviewed at latest revision, 4 unresolved discussions, all commit checks successful.

---

*[pkg/testutils/testcluster/testcluster.go, line 178 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjYfCg106LJZZjVsowR:-Kjdoi1-9WeTaKbeEgzk:b-w5l5vu) ([raw file](https://github.com/cockroachdb/cockroach/blob/fd569534c24b4e8d55f724901bc7c427ce242e8b/pkg/testutils/testcluster/testcluster.go#L178)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Oh, there is also the inconsistency with initialClusterStart - can you address that?
</blockquote></details>

got rid of `initialClusterStart`. PTAL.

---

*[pkg/testutils/testcluster/testcluster.go, line 183 at r7](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdZKLUeNO93DVWfEB-:-KjddjUjIqeSuuJ11hqj:bcdqs34) ([raw file](https://github.com/cockroachdb/cockroach/blob/3e8d6d3a460835e09537efa6cc46a2ea5c34dc03/pkg/testutils/testcluster/testcluster.go#L183)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Sadly, you need the `checkType` type annotation on both of these, otherwise you get an untyped constant. See https://play.golang.org/p/TsChFarg3g :(
</blockquote></details>

:( nice one

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 2 of 2 files at r8.
Review status: all files reviewed at latest revision, 6 unresolved discussions, all commit checks successful.

---

*[pkg/base/test_server_args.go, line 50 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdvWVj9S3mgLtDfvlG:-KjdvWVkXk_lHi7DY4Iu:b96b5pr) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/base/test_server_args.go#L50)):*
> is being added to a cluster

To a nonempty cluster, right?

---

*[pkg/testutils/testcluster/testcluster.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwSRYPfmOpdUDT1_G:-KjdwSRYPfmOpdUDT1_H:b-uev4hg) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L227)):*
> methods

method

---

*[pkg/testutils/testcluster/testcluster.go, line 233 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwYhsR4BmTq3IgnzD:-KjdwYhsR4BmTq3IgnzE:b187vup) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L233)):*
> filled it

filled in

---

*[pkg/testutils/testcluster/testcluster.go, line 242 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwbhFQeMOfrdRrou7:-KjdwbhFQeMOfrdRrou8:buuvvms) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L242)):*
> ```Go
> 		var stkCopy storage.StoreTestingKnobs
> 		if serverArgs.Knobs.Store == nil {
> 			stkCopy = storage.StoreTestingKnobs{}
> ```

this is essentially a no-op. how about
```
var stkCopy storage.StoreTestingKnobs
if stk := serverArgs.Knobs.Store; stk != nil {
  stkCopy = *stk.(*storage.StoreTestingKnobs)
}
```

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/testutils/testcluster/testcluster.go, line 227 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwSRYPfmOpdUDT1_G:-Kje4Qq_xxHxPrLyuRPj:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L227)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> methods

method
</blockquote></details>

Done.

---

*[pkg/testutils/testcluster/testcluster.go, line 233 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwYhsR4BmTq3IgnzD:-Kje4RC8N-aIB-lC530-:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L233)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> filled it

filled in
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 3 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[pkg/testutils/testcluster/testcluster.go, line 242 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdwbhFQeMOfrdRrou7:-Kje5RQSeVGRhS6bV3PN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/testutils/testcluster/testcluster.go#L242)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

this is essentially a no-op. how about
```
var stkCopy storage.StoreTestingKnobs
if stk := serverArgs.Knobs.Store; stk != nil {
  stkCopy = *stk.(*storage.StoreTestingKnobs)
}
```
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 1 of 1 files at r9.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/base/test_server_args.go, line 50 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdvWVj9S3mgLtDfvlG:-Kje5izi4BCAxlMA27jV:b9jd4mc) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/base/test_server_args.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

> is being added to a cluster

To a nonempty cluster, right?
</blockquote></details>

Did you mean to address this?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
Review status: 2 of 4 files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/base/test_server_args.go, line 50 at r8](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-KjdvWVj9S3mgLtDfvlG:-Kje6BymS7Gs9C8b9u8c:b4t4wyq) ([raw file](https://github.com/cockroachdb/cockroach/blob/4f7f7693a605976a884bff09608f8af91877a5c5/pkg/base/test_server_args.go#L50)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

Did you mean to address this?
</blockquote></details>

done now. Not sure how I missed this comment.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600)*
<!-- Sent from Reviewable.io -->

tamird commented:
Hurray! Thanks for bearing with me.

---

Reviewed 2 of 2 files at r10.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-Kje8qwS6UZ1loTNYHlH:b2b9dsj)*
<!-- Sent from Reviewable.io -->

andreimatei commented:
TFTR!

---

Review status: all files reviewed at latest revision, 1 unresolved discussion, all commit checks successful.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/9600#-:-KjeER6L8OowiSv6_XVU:b-814g8k)*
<!-- Sent from Reviewable.io -->
